### ISSUE TITLE
UserManager & SignInManager Oddities

### ISSUE THREAD

**Issue by Kizmar** [aspnet/Identity#934]
_See also: [my post on Stack Overflow about this](http://stackoverflow.com/questions/38876443/net-core-1-identity-usermanager-cached-user-list-not-updating/38877902?noredirect=1#comment65121735_38877902)_

We're noticing an issue when using these managers and haven't quite figured out the cause. 

_Note: the example code below was quickly typed up specifically for this post. It may not be fully complete, but should provide enough info to get what I'm talking about._

**What works:** 
UserManager<> and SignInManager<> work as expected when injected into a controller's constructor, and used only within that class.

**What does NOT seem to work:**
--1-- If we try to inject these managers into a single service layer class to encapsulate logic, the functions appear to be persisting changes to the database, but (I'm assuming) the changes aren't reflected in the managers in consequent calls. This is perplexing considering the DI scope should be per-request. In this case: a user would log in, change their password, log out, try to log back in, and would have to use the old password. (See the link above for more detail.)  _Example (this lives in a Service project separate from the WebAPI project):_
```
public interface IApplicationUserService {
	Task<IdentityResult> ChangePasswordAsync(string userName, string currentPassword, string newPassword);
}

public class ApplicationUserService {
	public ApplicationUserService(
		UserManager<ApplicationUser> userManager,
		SignInManager<ApplicationUser> signInManager) 
	{
		_userManager = userManager;
		_signInManager = signInManager;
	}

	private readonly UserManager<ApplicationUser> _userManager;
	private readonly SignInManager<ApplicationUser> _signInManager;

	public Task<IdentityResult> ChangePasswordAsync(string userName, string currentPassword, string newPassword) {
		var user = UserManager.FindByNameAsync(userName);		
		return _userManager.ChangePasswordAsync(user, currentPassword, newPassword);
	}
}

public class UserController {
	public UserController(IApplicationUserService applicationUserService) {
		_applicationUserService = applicationUserService;
	}

	private readonly IApplicationUserService _applicationUserService;

	[HttpPost("ChangePassword")]
	public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordViewModel model)
	{
		if (await _applicationUserService.ChangePasswordAsync(model.UserName, model.OldPassword, model.NewPassword) == IdentityResult.Success)
		{
			Logger.LogInformation($"User {user.FullName} changed their password successfully.");
			return Ok(ManageMessageId.ChangePasswordSuccess);
		}
		

		return BadRequest(ModelState);
	} 
}
```
--2-- If we try to pull these managers using IServiceProvider in a base controller, we get the same result as # 1. _Example:_
```
public class BaseController : Controller {
	public BaseController(IServiceProvider serviceProvider) {
		_userManager = serviceProvider.GetService<UserManager<ApplicationUser>>();
		_signInManager = serviceProvider.GetService<SignInManager<ApplicationUser>>();		
	}

	protected readonly UserManager<ApplicationUser> UserManager;
	protected readonly SignInManager<ApplicationUser> SignInManager;
}

public class UserController : BaseController {
	public UserController(IServiceProvider serviceProvider) : base(serviceProvider) {

	}

	[HttpPost("ChangePassword")]
	public async Task<IActionResult> ChangePassword([FromBody] ChangePasswordViewModel model)
	{
		var user = UserManager.FindByNameAsync(HttpContext.User.Identity.Name);
		if (user != null)
		{
			if (await UserManager.ChangePasswordAsync(user, model.OldPassword, model.NewPassword) == IdentityResult.Success)
			{
				Logger.LogInformation($"User {user.FullName} changed their password successfully.");
				return Ok(ManageMessageId.ChangePasswordSuccess);
			}
		}

		return BadRequest(ModelState);
	}
}
```
--3-- If we inject the managers into a controller's constructor, then pass to a base controller, we see the same result as # 1. (Example would be similar to # 2).

**Additional notes:**
We are using an ApplicationUser that inherits from IdentityUser that adds a few additional properties. Nothing fancy though.

In the service layer scenario, we wire up DI in `Startup.cs` like so: `services.AddScoped<IApplicationUserService, ApplicationUserService>();`.

We're also using JWT token middleware where we hit `PasswordSignInAsync(username, password, false, lockoutOnFailure: false);`. We've tried that with the service scenario and directly injecting the SignInManager. Similarly, we ended up having to inject the manager directly for it to recognize the new password. ...though it still only works when the managers are injected and used in a controller's class.

**In conclusion:**
We don't know if this is an issue with EF caching something, or the middleware caching something, or the DI not working as expected with the managers.

Are we going about this all wrong? Is it intentional that they only work when injected and used in a controller's class? If the managers are set up (by `app.UseIdentity()`) in the DI as per-request, I don't get why this is an issue.

**Comment by Kizmar** [comment]
I just pulled all of the token endpoint logic out of the middleware and created a LoginController that now houses it. I confirmed the managers work as expected now. The middleware must be somehow using a different copy of the managers that aren't being updated when changes are made.

**Comment by HaoK** [comment]
Yeah so middleware are going to pull from the wrong scope, so their lifetime will be singletons effectively for your app, that's why the behavior you are seeing is different vs controller lifetimes

**Comment by pierregorce** [comment]
Is there any solutions for use usermanager in the service layer AND in the controller ?
Can we define lifetime scoped for usermanger ?
